src/Common/PageCache.cpp

```diff
@@ -0,0 +1,655 @@
+
+namespace ProfileEvents
+{
+    extern const Event PageCacheChunkMisses;
+    extern const Event PageCacheChunkShared;
+    extern const Event PageCacheChunkDataHits;
+    extern const Event PageCacheChunkDataPartialHits;
+    extern const Event PageCacheChunkDataMisses;
+    extern const Event PageCacheBytesUnpinnedRoundedToPages;
+    extern const Event PageCacheBytesUnpinnedRoundedToHugePages;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int SYSTEM_ERROR;
+    extern const int MEMORY_LIMIT_EXCEEDED;
+    extern const int CANNOT_ALLOCATE_MEMORY;
+    extern const int INVALID_SETTING_VALUE;
+    extern const int FILE_DOESNT_EXIST;
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunknown-warning-option"
+#pragma clang diagnostic ignored "-Wreadability-make-member-function-const"
+
+PinnedPageChunk::PinnedPageChunk(PinnedPageChunk && c) noexcept
+    : cache(std::exchange(c.cache, nullptr)), chunk_(std::exchange(c.chunk_, nullptr)) {}
+
+PinnedPageChunk & PinnedPageChunk::operator=(PinnedPageChunk && c) noexcept
+{
+    if (cache)
+        cache->removeRef(chunk_);
+    cache = std::exchange(c.cache, nullptr);
+    chunk_ = std::exchange(c.chunk_, nullptr);
+    return *this;
+}
+
+PinnedPageChunk::~PinnedPageChunk() noexcept
+{
+    if (cache)
+        cache->removeRef(chunk_);
+}
+
+PinnedPageChunk::PinnedPageChunk(PageCache * _cache, PageChunk * chunk) noexcept : cache(_cache), chunk_(chunk) {}
+
+const PageChunk * PinnedPageChunk::chunk() const { return chunk_; }
+
+bool PinnedPageChunk::markPagePopulated(size_t page_idx)
+{
+    bool r = chunk_->pages_populated.set(page_idx);
+    return r;
+}
+
+void PinnedPageChunk::markPrefixPopulated(size_t bytes)
+{
+    for (size_t i = 0; i < (bytes + chunk_->page_size - 1) / chunk_->page_size; ++i)
+        markPagePopulated(i);
+}
+
+bool PinnedPageChunk::isPrefixPopulated(size_t bytes) const
+{
+    for (size_t i = 0; i < (bytes + chunk_->page_size - 1) / chunk_->page_size; ++i)
+        if (!chunk_->pages_populated.get(i))
+            return false;
+    return true;
+}
+
+AtomicBitSet::AtomicBitSet() = default;
+
+void AtomicBitSet::init(size_t nn)
+{
+    n = nn;
+    v = std::make_unique<std::atomic<UInt8>[]>((n + 7) / 8);
+}
+
+bool AtomicBitSet::get(size_t i) const
+{
+    return (v[i / 8] & (1 << (i % 8))) != 0;
+}
+
+bool AtomicBitSet::any() const
+{
+    for (size_t i = 0; i < (n + 7) / 8; ++i)
+        if (v[i])
+            return true;
+    return false;
+}
+
+bool AtomicBitSet::set(size_t i)
+{
+    UInt8 prev = v[i / 8].fetch_or(1 << (i % 8));
+    return (prev & (1 << (i % 8))) == 0;
+}
+
+bool AtomicBitSet::set(size_t i, bool val)
+{
+    if (val)
+        return set(i);
+    else
+        return unset(i);
+}
+
+bool AtomicBitSet::unset(size_t i)
+{
+    UInt8 prev = v[i / 8].fetch_and(~(1 << (i % 8)));
+    return (prev & (1 << (i % 8))) != 0;
+}
+
+void AtomicBitSet::unsetAll()
+{
+    memset(&v[0], 0, (n + 7) / 8);
```

Comment:
Hm, it surely is per-byte atomic on x86, but I'm not entirely sure whether some theoretical UB compiler shenanigans are possible. Replaced with a loop (with memory_order_relaxed) just in case, so that the next person to read the code doesn't waste time wondering about it.

There are no memory barriers, so this memset has the same guarantees as memory_order_relaxed, but that's sufficient here. (And would be sufficient for all other AtomicBitSet operations, I think - we're not using the bitset to synchronize access to other data, there's always a mutex involved; but not super sure about it at this point, it's been a while, and this code turned out pretty complicated, unfortunately).